src/debug/ee/controller.cpp

```diff
@@ -7938,15 +7894,24 @@ void DebuggerJMCStepper::TriggerMethodEnter(Thread * thread,
     if (!fIsUserCode)
         return;
 
+    if ((m_eMode != cStepIn) && (m_fp != LEAF_MOST_FRAME))
+    {
+        // We can run a stack trace in managed code (eg, not a stub)
+        ControllerStackInfo info;
+        StackTraceTicket ticket(ip);
+        info.GetStackInfo(ticket, thread, LEAF_MOST_FRAME, NULL);
+
+        if (IsCloserToLeaf(info.m_activeFrame.fp, m_fp))
+        {
+            LOG((LF_CORDB, LL_INFO100000, "DJMCStepper::TME, skipping b/c step-out fp\n"));
+            return;
+        }
+    }
+
     // MethodEnter is only enabled when we want to stop in a JMC function.
     // And that's where we are now. So patch the ip and resume.
     // The stepper will hit the patch, and stop.
 
-    // It's a good thing we have the fp passed in, because we have no other
-    // way of getting it. We can't do a stack trace here (the stack trace
-    // would start at the last pushed Frame, which miss a lot of managed
-    // frames).
```

Comment:
This comment was currently nonsense. 
- The fp passed in is always set to LEAF_MOST_FRAME which is `null`.  
- We can get the fp in at least some cases as shown above.